File: leetcode/medium/arrays_and_strings/next_permutation.cc
#include <algorithm>

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if(nums.size()<2){
            return;
        }
        int i;
        for (i = nums.size() - 1; i > 0 && nums[i - 1] >= nums[i]; i--);
        if (i == 0) {
            std::sort(nums.begin(), nums.end());
            return;
        }
        std::sort(nums.begin()+i, nums.end());
        int j;
        for (j = i; j < nums.size() && nums[i - 1] >= nums[j]; j++);
        std::swap(nums[i-1], nums[j]);
    }
};

File: leetcode/medium/arrays_and_strings/NextPermutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        if(nums.length<2){
            return;
        }
        int i;
        for (i = nums.length - 1; i > 0 && nums[i - 1] >= nums[i]; i--);
        if (i == 0) {
            Arrays.sort(nums);
            return;
        }
        int[] aux = Arrays.copyOfRange(nums, i, nums.length);
        Arrays.sort(aux);
        for (int k = 0; k < aux.length; k++) {
            nums[i + k] = aux[k];
        }
        int j;
        for (j = i; j < nums.length && nums[i - 1] >= nums[j]; j++);
        //swap nums[i-1] and nums[j]
        nums[i-1]=nums[i-1]+nums[j];
        nums[j]=nums[i-1]-nums[j];
        nums[i-1]=nums[i-1]-nums[j];
    }
}